### DISCUSSION TITLE
Error 404 on code-server on iPad via Raspberry Pi

### DISCUSSION THREAD

**Discussion by LJ-03** [coder/code-server#4934] [Q&A] [answered]
Hi there! So I set up a raspberry Pi 4b to wirelessly connect to my iPad, so I can ssh into the Pi from my iPad, using Rob from TechCraft’s wireless tutorial. I then followed a guide to set up code-server on my Pi so that I can view it in browser on my iPad too. I followed the instructions in both of these guides:
https://medium.com/swlh/vs-code-for-ipad-and-other-phones-and-tablets-252e8cace2aa
https://github.com/andreivisan/Raspberry-Pi-4-Code-Server

Everything seemed to be fine, and I’ve been able to get code-server running via `sudo systemctl start code-server`, but when I route to my iPad’s IP address followed by the chosen port (8080), it gives me a 404 error.

I’ll go through what I have for my setup. 

I’m running Node v 14.19.0 (I had a command line prompt that this was the latest compatible version with code-server)
npm v6.14.16
Yarn 1.22.17

I had yarn already installed before I ran the guide, so I believe it installed via Yarn when I ran `curl -fsSL https://code-server.dev/install.sh | sh` to install code-server. After that I was prompted to set my path via this command: `PATH="$(yarn global bin):$PATH"` which I did. For my code-server `.config.yaml` and `code-server.service` files, I created these with the following commands:

`sudo vim /lib/systemd/system/code-server.service`
Content:
```
[Unit]
Description=code-server
After=network.target

[Service]
Type=simple
Environment=PASSWORD=my_password
ExecStart=/home/pi/.yarn/bin/code-server —bind-addr 0.0.0.0:8080 —user-data-dir /var/lib/code-server —auth password
Restart=always

[Install]
WantedBy=multi-user.target

```
`sudo vim /home/pi/.config/code-server/config.yaml`
Content:
```
bind-addr: 0.0.0.0:8080
auth: password
password: my_password
cert: false
```

After setting and saving these, I used the standard:
`sudo daemon-reload`
`sudo systemctl start code-server`
`sudo systemctl enable code-server`
`sudo systemctl status code-server`

Which then gives me the following:

```
● code-server.service - code-server
     Loaded: loaded (/lib/systemd/system/code-server.service; enabled; vendor preset: enabled)
     Active: active (running) since Wed 2022-03-02 17:21:59 AEST; 3min ago
   Main PID: 2847 (node)
      Tasks: 33 (limit: 4915)
        CPU: 10.876s

  CGroup: /system.slice/code-server.service
             ├─2847 node /home/pi/.yarn/bin/code-server --bind-addr 0.0.0.0:8080 --user-data-dir /var/lib/code-server --auth password
             ├─2859 /usr/local/bin/node /home/pi/.yarn/bin/code-server --bind-addr 0.0.0.0:8080 --user-data-dir /var/lib/code-server --auth password
             └─2885 /usr/local/bin/node /home/pi/.config/yarn/global/node_modules/code-server/vendor/modules/code-oss-dev/out/bootstrap-fork --type=ptyHost

Mar 02 17:22:09 raspberrypi code-server[2847]:   phase: 'loading',
Mar 02 17:22:09 raspberrypi code-server[2847]:   moduleId: 'spdlog',
Mar 02 17:22:09 raspberrypi code-server[2847]:   neededBy: [ '===anonymous3===' ]
Mar 02 17:22:09 raspberrypi code-server[2847]: }
Mar 02 17:22:19 raspberrypi code-server[2847]: [17:22:19] [Error: ENOENT: no such file or directory, scandir '/var/lib/code-server/logs'] {
Mar 02 17:22:19 raspberrypi code-server[2847]:   errno: -2,
Mar 02 17:22:19 raspberrypi code-server[2847]:   code: 'ENOENT',
Mar 02 17:22:19 raspberrypi code-server[2847]:   syscall: 'scandir',
Mar 02 17:22:19 raspberrypi code-server[2847]:   path: '/var/lib/code-server/logs’
Mar 02 17:22:19 raspberrypi code-server[2847]: }

```
And that’s about it! I’ve tried re-installing code-server a few times, looked at a few different configurations of the `code-server.service` and `config.yaml` files but no such luck. Any insight would be greatly appreciated!!

EDIT: I don’t know if it’s worth mentioning that because of the way my Pi is set up that I have two IPs when I run `hostname -I`. I’ve tried both when running code-server and both give the 404 error.

**Comment by lg-nineoneninethreenine** [comment]
Absolutely highest possibility - https://github.com/coder/code-server/issues/4900

But I thought it was fixed - I need more details, like are you using the brand new Raspberry Pi OS 64-bit or the 32-bit one? Or are you using the Legacy edition?
Moreover, I need the code-server version, your iPadOS version, some code-server's built-in error logs(I think you might know, but if you don't, it's `code-server --log debug`) as well as DevTools Console screenshots. I'll try repros once you send me the required data during the weekend(I'm a teen and I need the RasPi during that time), but if there are simple fixes, @jsjoeio might point it out.
Moreover, are there any peripherals that could possibly disrupt the performance?

**Reply by LJ-03** [reply]
Oh wow @im-coder-lg , just had a very brief skim of #4900, seems like it could be pretty involved. It’s late here, so I’ll aim to get those details to you tomorrow when I can 😁 Thanks for reaching out!

**Reply by lg-nineoneninethreenine** [reply]
No worries ☺️ anyway it's evening here, send the data later when you are free. My only aim as of now is trying repro steps (and doing my homework lol) and getting more insight on this, then trying to fix it ASAP. But take your rest and then send that data when you can!

**Reply by LJ-03** [reply]
Okay so the extra info you need:
Running arm v7l, so 32 bit of Pi-OS
Code-server: v4.0.2
IPadOS: v15.3.1
After ssh’ing to my Pi, running code-server with error logs, I get the following output:

```
[2022-03-03T07:19:37.242Z] info  code-server 4.0.2 5cdfe74686aa73e023f8354a9a6014eb30caa7dd
[2022-03-03T07:19:37.245Z] info  Using user-data-dir ~/.local/share/code-server
[2022-03-03T07:19:37.283Z] info  Using config file ~/.config/code-server/config.yaml
[2022-03-03T07:19:37.283Z] info  HTTP server listening on http://0.0.0.0:8080/ 
[2022-03-03T07:19:37.284Z] info    - Authentication is enabled
[2022-03-03T07:19:37.284Z] info      - Using password from ~/.config/code-server/config.yaml
[2022-03-03T07:19:37.284Z] info    - Not serving HTTPS 
[17:19:46] Extension host agent started.
[2022-03-03T07:19:46.456Z] debug redirecting from / to ./login
[17:19:47] [Theme Service] Scanning for theme extension...
[17:19:47] [Theme Service] theme-abyss
[17:19:47] [Theme Service] theme-defaults
[17:19:47] [Theme Service] theme-kimbie-dark
[17:19:47] [Theme Service] theme-monokai
[17:19:47] [Theme Service] theme-monokai-dimmed
[17:19:47] [Theme Service] theme-quietlight
[17:19:47] [Theme Service] theme-red
[17:19:47] [Theme Service] theme-solarized-dark
[17:19:47] [Theme Service] theme-solarized-light
[17:19:47] [Theme Service] theme-tomorrow-night-blue
[2022-03-03T07:19:50.783Z] debug no valid cookie doman {"host":"MY_IP_ADDRESS"}
[2022-03-03T07:19:50.786Z] debug redirecting from /login to ./
[2022-03-03T07:20:45.800Z] debug 0 active connections
```

In terms of dev tools console, I’m running safari on the iPad, and it doesn’t seem to have the dev tools extension available. Do I need to run Chrome for iPad and get dev tools that way for the screenshots?

And lastly, with peripherals, I really just have the Pi wirelessly connected over my home network to the iPad. I use the iPad magic keyboard, AirPods, and the Apple Pencil is connected but never used while i use code-server, but I don’t see how either of those would affect anything. If you’re thinking of any other potential problem devices that may cause issues, let me know. I have things connected via Bluetooth like AirPods, but no other dedicated peripherals.

And no stress with the time! You’ve gotta get homework done! Quite frankly I’m used to waiting days or more for sub-par help on stack overflow so this is already a massive help 😂 thanks heaps again! And let me know if you need any other info from me 😁

**Reply by lg-nineoneninethreenine** [reply]
> In terms of dev tools console, I’m running safari on the iPad, and it doesn’t seem to have the dev tools extension available. Do I need to run Chrome for iPad and get dev tools that way for the screenshots?

Oh, I forgot, can you try getting it from your PC/Mac? Use Google Chrome btw

> And lastly, with peripherals, I really just have the Pi wirelessly connected over my home network to the iPad. I use the iPad magic keyboard, AirPods, and the Apple Pencil is connected but never used while i use code-server, but I don’t see how either of those would affect anything. If you’re thinking of any other potential problem devices that may cause issues, let me know. I have things connected via Bluetooth like AirPods, but no other dedicated peripherals.

Mostly, it might have caused something if it was a USB plugged into the Pi, but it's all my imagination. Just ignore that, anyway I expected no disturbances from it.

> And no stress with the time! You’ve gotta get homework done! Quite frankly I’m used to waiting days or more for sub-par help on stack overflow so this is already a massive help 😂 thanks heaps again! And let me know if you need any other info from me 😁

No worries, anyway I have less time to contribute here or anywhere, so all I have to do is utilise it usefully. You kinda made my day :)

For the error logs on DevTools, maybe you could try using Chrome on a real computer or maybe connect your Raspberry Pi to your TV or monitor, get some GUI thingies and test it on the Pi itself. Kinda reliable.

**Reply by LJ-03** [reply]
> Oh, I forgot, can you try getting it from your PC/Mac? Use Google Chrome btw

Sure! From my Mac in Chrome I’m getting 2 errors:
`GET http://MY_IP_ADDRESS:8080/ 404 (Not Found)`
And 
`GET http://MY_IP_ADDRESS:8080/manifest.json 404 (Not Found)`
I’m getting nothing else; no warnings, no info etc. Just those two errors. But if I’ve missed some key detail from the dev tools UI just let me know and I can get that for you again.

> No worries, anyway I have less time to contribute here or anywhere, so all I have to do is utilise it usefully. You kinda made my day :)

Always glad to make someone’s day!

**Reply by lg-nineoneninethreenine** [reply]
Possibility - missing 404 page?

**Reply by LJ-03** [reply]
I'm at work at the moment so I can't ssh back into the Pi at the moment, but I saw that it was pointing to a 404 page, which I had a quick look and could see some code for a basic 404 page.

**Reply by taeuscherpferd** [reply]
Just chiming in here, I have almost the exact same setup (raspberry pi 4, ssh with ipad), and I'm getting the exact same error. Something that was different in my case was that a bunch of packages failed to install as well. I had to go and install Yauzl, yazl, xterm-headless, spdlog, and minimist globally separately before I was able to get the server to start, which immediately routed me to a 404 page after logging in

**Reply by LJ-03** [reply]
Interesting - so you haven't managed to resolve your issue yet either, @taeuscherpferd ?

**Reply by taeuscherpferd** [reply]
Correct. Also the version that I'm trying to use is 4.0.2

**Reply by lg-nineoneninethreenine** [reply]
@jsjoeio is this really fixed?

**Reply by taeuscherpferd** [reply]
I just managed to get it working with npm i code-server@4.1.0-4946-9e620e90f53fb91338a2ba1aaa2e556d42ae52d5. Seeing as 4.1.0 was released yesterday, I would guess that the fix is coming in that version. Hopefully there's a proper 4.1.0 release build pushed to npm soon :)

**Reply by jsjoeio** [reply]
@taeuscherpferd yes! We had a hiccup but should have it out soon

**Reply by lg-nineoneninethreenine** [reply]
Phew this is almost solved!!! Why is the issue closed then?

**Reply by jsjoeio** [reply]
It's out on npm now!

**Reply by lg-nineoneninethreenine** [reply]
![yeahboiiii](https://c.tenor.com/nRjZhsODQ-MAAAAM/yeah-boy-yes.gif)

**Reply by LJ-03** [reply]
So @taeuscherpferd what did you do? Just update npm? Would I just need to update npm and it should be okay? Even if I installed code-server via yarn?

**Reply by lg-nineoneninethreenine** [reply]
I guess all you have to do is just get the beta package from npm or yarn.

**Comment by LJ-03** [comment, marked as answer]
I just upgraded code-server to version 4.10 just as @taeuscherpferd and am happy to report that it is working! 🥳🎉 Thank you to everyone who was so quick to jump in and help! This is an awesome and _very_ helpful community!!

**Reply by ravikg** [reply]
I was also having same issues. And upgrading `code-server` to latest version `4.10` solved the issue.

To upgrade, just re-ran the install command:
```bash
$ curl -fsSL https://code-server.dev/install.sh | sh
```

**Reply by lg-nineoneninethreenine** [reply]
Happy to find it's solved!

**Reply by jsjoeio** [reply]
Hooray! Very happy to hear that!
